docs/metrics.md

```diff
@@ -46,4 +46,136 @@ metrics, to prevent expanding the required set of dependencies. Please see [armo
 
 ### Targeted vs. Untargeted Attacks
 
-For targeted attacks, each metric will be reported twice for adversarial data: once relative to the ground truth labels and once relative to the target labels.  For untargeted attacks, each metric is only reported relative to the ground truth labels.  Performance relative to ground truth measures the effectiveness of the defense, indicating the ability of the model to make correct predictions despite the perturbed input.  Performance relative to target labels measures the effectiveness of the attack, indicating the ability of the attacker to force the model to make predictions that are not only incorrect, but that align with the attackers chosen output.
\ No newline at end of file
+For targeted attacks, each metric will be reported twice for adversarial data: once relative to the ground truth labels and once relative to the target labels.  For untargeted attacks, each metric is only reported relative to the ground truth labels.  Performance relative to ground truth measures the effectiveness of the defense, indicating the ability of the model to make correct predictions despite the perturbed input.  Performance relative to target labels measures the effectiveness of the attack, indicating the ability of the attacker to force the model to make predictions that are not only incorrect, but that align with the attackers chosen output.
+
+# Instrumentation
+
+The `armory.metrics.instrument` module implements functionality to flexibly capture values for measurement.
+
+## Usage
+
+The primary mechanisms are largely based off of the logging paradigm of loggers and handlers.
+Loggers here are referred to as probes, and handlers are referred to as meters.
+
+To capture values in-line, use probe:
+```
+# Module imports section
+from armory.metrics import instrument
+probe = instrument.get_probe()
+
+# ...
+
+# In the code
+probe.update(name=value)
+```
+
+However, this will fall on the floor unless a meter is connected to record values.
+
+
+### Preprocessing
+
+Probes can perform preprocessing of the updating values.
+This involves specifying a function or sequence of functions to perform when a value is updated.
+However, this is done lazily - it is only performed if a meter is connected and is measuring.
+
+For instance, to extract a pytorch tensor `layer_3_output`, you could do:
+```
+probe.update(lambda x: x.detach().cpu().numpy(), a=layer_3_output)
+```
+Or in a longer form:
+```
+probe.update(lambda x: x.detach(), lambda x: x.cpu().numpy(), a=layer_3_output)
```

Comment:
one thing that helped my understanding but wasn't immediately clear is that when you call

```
probe.update(arbitrary_name=value)
```

you are defining `arbitrary_name` as a name that can then be used when passing a string `"probe_name.arbitrary_name"` to `add_meter()`

Reply:
Added in docs. Need to update them further once done.